api/utils/pkixname/parser.go

```diff
@@ -0,0 +1,583 @@
+				return nil, fmt.Errorf("hexstring not supported: %s", errTrace(pos))
+			case '"':
+				state = tokenizeStateStringQuote
+				// Note that a quoted string may be empty.
+				continue
+			default:
+				state = tokenizeStateString
+				// Rune not consumed.
+			}
+
+		case tokenizeStateStringEnd:
+			switch r {
+			case ' ':
+				trailingSpaceBuf.WriteRune(r)
+				continue
+			case '+', ',', ';':
+				trailingSpaceBuf.Reset() // whitespace discarded.
+				emitBuffer(tokenString)
+				transitionToNameComponent(r)
+				continue
+			default:
+				trailingSpaceBuf.WriteTo(buf) // whitespace copied back.
+				trailingSpaceBuf.Reset()
+				state = tokenizeStateString
+				// Rune not consumed.
+			}
+		}
+
+		switch state {
+		case tokenizeStateInit, tokenizeStateNameComponent:
+			switch {
+			case isAttrType(r):
+				state = tokenizeStateAttrType
+				buf.WriteRune(r)
+			default:
+				return nil, fmt.Errorf("want attributeType, found %q: %s", r, errTrace(pos))
+			}
+
+		case tokenizeStateAttrType:
+			switch {
+			case isAttrType(r):
+				buf.WriteRune(r)
+			case r == '=':
+				emitBuffer(tokenAttrType)
+				emit(tokenEqual)
+				state = tokenizeStateStringStart
+			default:
+				return nil, fmt.Errorf("want attributeType or '=', found %q: %s", r, errTrace(pos))
+			}
+
+		case tokenizeStateString:
+			switch r {
+			case '+', ',', ';':
+				emitBuffer(tokenString)
+				transitionToNameComponent(r)
+			case '\\':
+				escapeStart()
+			case '<', '>', '"':
+				// We could '<' and '>', but let's be strict.
+				return nil, fmt.Errorf("special character %q not quoted: %s", r, errTrace(pos))
+			case ' ':
+				trailingSpaceBuf.WriteRune(r)
+				state = tokenizeStateStringEnd
+			case '=', '#': // Go does this.
+				// NOT OK per RFC, should be escaped.
+				fallthrough
+			default:
+				buf.WriteRune(r)
+			}
+
+		case tokenizeStateStringEscape:
+			switch r {
+			case ' ': // Go does this.
+				// OK per RFC and allows the "\\ " trick.
+				fallthrough
+			case ',', '=', '+', '<', '>', '#', ';', '\\', '"':
+				buf.WriteRune(r)
+				escapeEnd()
+			default:
+				return nil, fmt.Errorf("unexpected escaped character %q: %s", r, errTrace(pos))
+			}
+
+		case tokenizeStateStringQuote:
+			switch r {
+			case '\\':
+				escapeStart()
+			case '"':
+				emitBuffer(tokenString)
+				state = tokenizeStateStringQuoteEnd
+			default:
+				buf.WriteRune(r)
+			}
+
+		case tokenizeStateStringQuoteEnd:
+			switch r {
+			case '+', ',', ';':
+				transitionToNameComponent(r)
+			default:
+				return nil, fmt.Errorf("want '+' or ',', found %q: %s", r, errTrace(pos))
+			}
+		}
+	}
+
+	// Input ended, check the final state.
+	switch state {
+	case tokenizeStateInit:
+		// OK.
+	case tokenizeStateNameComponent:
+		return nil, fmt.Errorf("want attributeType, found EOF")
+	case tokenizeStateAttrType:
+		return nil, fmt.Errorf("want attributeType or '=', found EOF")
+	case tokenizeStateStringStart:
+		return nil, fmt.Errorf("want attributeValue, found EOF")
```

Comment:
I think the RFC technically allows for empty values, if I've read it correctly? It's definitely reasonable to reject that case but it might be worth a note, assuming my reading comprehension checks out.